// dynamicsnode/src/CRMDataTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EntityReference = (function () {
    function EntityReference(Id, LogicalName) {
        this.Id = Id;
        this.LogicalName = LogicalName;
        this.__typeName = "Microsoft.Xrm.Sdk,Microsoft.Xrm.Sdk.EntityReference";
    }
    return EntityReference;
}());
exports.EntityReference = EntityReference;
var OptionSetValue = (function () {
    function OptionSetValue(Value) {
        this.Value = Value;
        this.__typeName = "Microsoft.Xrm.Sdk,Microsoft.Xrm.Sdk.OptionSetValue";
    }
    return OptionSetValue;
}());
exports.OptionSetValue = OptionSetValue;
var Entity = (function () {
    function Entity() {
        this.__typeName = "Microsoft.Xrm.Sdk,Microsoft.Xrm.Sdk.Entity";
    }
    return Entity;
}());
exports.Entity = Entity;
var EntityMetadata = (function () {
    function EntityMetadata() {
    }
    return EntityMetadata;
}());
exports.EntityMetadata = EntityMetadata;
var LocalizedLabel = (function () {
    function LocalizedLabel() {
    }
    return LocalizedLabel;
}());
exports.LocalizedLabel = LocalizedLabel;
var Label = (function () {
    function Label() {
    }
    return Label;
}());
exports.Label = Label;
var OptionMetadata = (function () {
    function OptionMetadata() {
    }
    return OptionMetadata;
}());
exports.OptionMetadata = OptionMetadata;
var OptionsetMetadata = (function () {
    function OptionsetMetadata() {
    }
    return OptionsetMetadata;
}());
exports.OptionsetMetadata = OptionsetMetadata;
var BooleanOptionsetMetadata = (function () {
    function BooleanOptionsetMetadata() {
    }
    return BooleanOptionsetMetadata;
}());
exports.BooleanOptionsetMetadata = BooleanOptionsetMetadata;
var AttributeMetadata = (function () {
    function AttributeMetadata() {
    }
    return AttributeMetadata;
}());
exports.AttributeMetadata = AttributeMetadata;
var AttributeTypeCode;
(function (AttributeTypeCode) {
    AttributeTypeCode[AttributeTypeCode["BigInt"] = 18] = "BigInt";
    AttributeTypeCode[AttributeTypeCode["Boolean"] = 0] = "Boolean";
    AttributeTypeCode[AttributeTypeCode["CalendarRules"] = 16] = "CalendarRules";
    AttributeTypeCode[AttributeTypeCode["Customer"] = 1] = "Customer";
    AttributeTypeCode[AttributeTypeCode["DateTime"] = 2] = "DateTime";
    AttributeTypeCode[AttributeTypeCode["Decimal"] = 3] = "Decimal";
    AttributeTypeCode[AttributeTypeCode["Double"] = 4] = "Double";
    AttributeTypeCode[AttributeTypeCode["EntityName"] = 20] = "EntityName";
    AttributeTypeCode[AttributeTypeCode["Integer"] = 5] = "Integer";
    AttributeTypeCode[AttributeTypeCode["Lookup"] = 6] = "Lookup";
    AttributeTypeCode[AttributeTypeCode["ManagedProperty"] = 19] = "ManagedProperty";
    AttributeTypeCode[AttributeTypeCode["Memo"] = 7] = "Memo";
    AttributeTypeCode[AttributeTypeCode["Money"] = 8] = "Money";
    AttributeTypeCode[AttributeTypeCode["Owner"] = 9] = "Owner";
    AttributeTypeCode[AttributeTypeCode["PartyList"] = 10] = "PartyList";
    AttributeTypeCode[AttributeTypeCode["Picklist"] = 11] = "Picklist";
    AttributeTypeCode[AttributeTypeCode["State"] = 12] = "State";
    AttributeTypeCode[AttributeTypeCode["Status"] = 13] = "Status";
    AttributeTypeCode[AttributeTypeCode["String"] = 14] = "String";
    AttributeTypeCode[AttributeTypeCode["Uniqueidentifier"] = 15] = "Uniqueidentifier";
    AttributeTypeCode[AttributeTypeCode["Virtual"] = 17] = "Virtual";
})(AttributeTypeCode = exports.AttributeTypeCode || (exports.AttributeTypeCode = {}));
var EntityFilters;
(function (EntityFilters) {
    EntityFilters[EntityFilters["Entity"] = 1] = "Entity";
    EntityFilters[EntityFilters["Default"] = 1] = "Default";
    EntityFilters[EntityFilters["Attributes"] = 2] = "Attributes";
    EntityFilters[EntityFilters["Privileges"] = 4] = "Privileges";
    EntityFilters[EntityFilters["Relationships"] = 8] = "Relationships";
    EntityFilters[EntityFilters["All"] = 15] = "All";
})(EntityFilters = exports.EntityFilters || (exports.EntityFilters = {}));
