// dynamicsnode/src/Messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var WhoAmIRequest = (function () {
    function WhoAmIRequest() {
        this.__typeName = "Microsoft.Crm.Sdk.Proxy,Microsoft.Crm.Sdk.Messages.WhoAmIRequest";
    }
    return WhoAmIRequest;
}());
exports.WhoAmIRequest = WhoAmIRequest;
var WhoAmIResponse = (function () {
    function WhoAmIResponse() {
    }
    return WhoAmIResponse;
}());
exports.WhoAmIResponse = WhoAmIResponse;
var AssignRequest = (function () {
    function AssignRequest() {
        this.__typeName = "Microsoft.Crm.Sdk.Proxy,Microsoft.Crm.Sdk.Messages.AssignRequest";
    }
    return AssignRequest;
}());
exports.AssignRequest = AssignRequest;
var AssignResponse = (function () {
    function AssignResponse() {
    }
    return AssignResponse;
}());
exports.AssignResponse = AssignResponse;
var RetrieveEntityRequest = (function () {
    function RetrieveEntityRequest(LogicalName, EntityFilters) {
        this.LogicalName = LogicalName;
        this.EntityFilters = EntityFilters;
        this.__typeName = "Microsoft.Xrm.Sdk,Microsoft.Xrm.Sdk.Messages.RetrieveEntityRequest";
    }
    return RetrieveEntityRequest;
}());
exports.RetrieveEntityRequest = RetrieveEntityRequest;
var RetrieveEntityResponse = (function () {
    function RetrieveEntityResponse() {
    }
    return RetrieveEntityResponse;
}());
exports.RetrieveEntityResponse = RetrieveEntityResponse;
var SetStateRequest = (function () {
    function SetStateRequest() {
        this.__typeName = "Microsoft.Crm.Sdk.Proxy,Microsoft.Crm.Sdk.Messages.SetStateRequest";
    }
    return SetStateRequest;
}());
exports.SetStateRequest = SetStateRequest;
var SetBusinessSystemUserRequest = (function () {
    function SetBusinessSystemUserRequest() {
        this.__typeName = "Microsoft.Crm.Sdk.Proxy,Microsoft.Crm.Sdk.Messages.SetBusinessSystemUserRequest";
    }
    return SetBusinessSystemUserRequest;
}());
exports.SetBusinessSystemUserRequest = SetBusinessSystemUserRequest;
